scrapy_poet/page_input_providers.py

```diff
@@ -306,32 +310,50 @@ async def __call__(
-                )
-                # RecursionError NOT raised when ``AsyncioSelectorReactor`` is used.
-                # Could be related: https://github.com/python/cpython/issues/93837
-
-                # Need to check before awaiting on the ``asyncio.Future``
-                # before it gets stuck on a potential deadlock.
-                if asyncio.isfuture(deferred_or_future):
-                    if self.check_if_deadlock(request):
-                        raise ProviderDependencyDeadlockError(
-                            self.template_deadlock_msg.format(plan=plan)
+            # If dependencies are already built for the PO, we can build it here.
+            # Avoid calling build_instances() since it may result in deadlocks.
+            po_deps = [cls for cls, kwargs_spec in plan.dependencies][:-1]
+            if prev_instances.keys() == set(po_deps):
+                _, kwargs_spec = plan.dependencies[-1]
+                po_instances = {
+                    page_object_cls: page_object_cls(
+                        **kwargs_spec.kwargs(prev_instances)
+                    )
+                }
+
+            else:
+                try:
+                    deferred_or_future = maybe_deferred_to_future(
+                        self.injector.build_instances(
+                            request, response, plan, prev_instances
                         )
-
-                po_instances = await deferred_or_future
-            except RecursionError:
-                raise ProviderDependencyDeadlockError(
-                    self.template_deadlock_msg.format(plan=plan)
-                )
-
-            page_object = po_instances[page_object_cls]
-            item = await page_object.to_item()
-
-            self.update_cache(request, po_instances)
-            self.update_cache(request, {type(item): item})
+                    )
+                    # RecursionError NOT raised when ``AsyncioSelectorReactor`` is used.
+                    # Could be related: https://github.com/python/cpython/issues/93837
+
+                    # Need to check before awaiting on the ``asyncio.Future``
+                    # before it gets stuck on a potential deadlock.
+                    if asyncio.isfuture(deferred_or_future):
+                        if self.check_if_deadlock(request):
+                            raise ProviderDependencyDeadlockError(
+                                self.template_deadlock_msg.format(plan=plan)
+                            )
+
+                    po_instances = await deferred_or_future
+                except RecursionError:
+                    raise ProviderDependencyDeadlockError(
+                        self.template_deadlock_msg.format(plan=plan)
+                    )
+
+            # See if the recursive calls produced an item and put it in the cache.
+            # Otherwise, ``to_item()`` will be called multiple times which can
+            # lead to bad values in some fields.
```

Comment:
TODO: There's probably a better way to detect and prevent items from recursive calls from bubbling up.

Refactor this when the other test cases are already in place.

Reply:
Done in https://github.com/scrapinghub/scrapy-poet/pull/164/commits/88cdc5e47db35bdaf21a45b5d23eae17d8efa6cd